### PR TITLE
Raft stability

### PR DESCRIPTION
**Pull request by benbjohnson**
## Overview

This pull request adds several `raft` package clean up fixes.

A list of high level changes:

- Refactor "applied index" into the `FSM` only - previously this index was shared between the `raft.Log` and the `FSM`.

- Refactor `FSM` API to implement `io.WriterTo` and `io.ReaderFrom`. The "applied index" refactor above altered the API so conforming to a standard API makes this easier to test.

- Fix resnapshotting in some edge cases scenarios - the tradeoff of the temporary in-memory staging area for Raft entries requires that the snapshot be refetched in some election scenarios.

- Fix race detection on state loop startup.

- Fix race condition when setting the incoming stream.

- Fix edge cases in log truncation & trimming.

- Fix new term notification in follower & candidate states.

This pull request also documents the various internal fields in `raft.Log`.

### CODE REVIEW

**Comment by pauldix** [comment]
Just did a quick review and LGTM. Although I haven't fully digested it yet, we need to get this in for RC17 so I'll merge and hopefully @otoolep, @corylanou, and @dgnorton can do further review tomorrow.

File: messaging/broker.go [hunk 1]
```diff
@@ -352,32 +351,32 @@ func copyFileN(w io.Writer, path string, n int64) (int64, error) {
 	return io.CopyN(w, f, n)
 }
 
-// Restore reads the broker state.
-func (b *Broker) Restore(r io.Reader) error {
+// ReadFrom reads a broker snapshot from r.
```
Comment (otoolep) [inline messaging/broker.go]: Minor: describe what the returned `int64` is?

File: raft/errors.go [hunk 2]
```diff
@@ -28,10 +28,6 @@ var (
 	// ErrOutOfDateLog is returned when a candidate's log is not up to date.
 	ErrOutOfDateLog = errors.New("out of date log")
 
-	// ErrUncommittedIndex is returned when a stream is started from an
```
Comment (otoolep) [inline raft/errors.go]: This can't happen anymore?

File: raft/log.go [hunk 3]
```diff
@@ -25,19 +25,15 @@ import (
 // FSM represents the state machine that the log is applied to.
 // The FSM must maintain the highest index that it has seen.
 type FSM interface {
+	io.WriterTo
```
Comment (otoolep) [inline raft/log.go]: Worth documenting what is sent over these interfaces? I think so.

File: raft/log.go [hunk 4]
```diff
@@ -436,6 +523,12 @@ func (l *Log) writeTerm(term uint64) error {
 	return ioutil.WriteFile(l.termPath(), b, 0666)
 }
 
+// setTerm sets the current term if it's higher.
```
Comment (otoolep) [inline raft/log.go]: This comment does not look correct. I don't see any check within the code. It just sets the term unconditionally.

File: raft/log.go [hunk 5]
```diff
@@ -1095,18 +1220,18 @@ func (l *Log) internalApply(typ LogEntryType, command []byte) (index uint64, err
 }
 
 // Wait blocks until a given index is applied.
-func (l *Log) Wait(index uint64) error {
+func (l *Log) Wait(idx uint64) error {
 	// TODO(benbjohnson): Check for leadership change (?).
```
Comment (otoolep) [inline raft/log.go]: BTW, should we open tickets for these two items?

File: raft/log.go [hunk 6]
```diff
@@ -1145,9 +1270,27 @@ func (l *Log) waitUncommitted(index uint64) error {
...
 // append adds a log entry to the list of entries.
-func (l *Log) append(e *LogEntry) {
-	//l.tracef("append: idx=%d, prev=%d", e.Index, l.lastLogIndex)
-	assert(e.Index == l.lastLogIndex+1, "non-contiguous log index(%d): idx=%d, prev=%d", l.id, e.Index, l.lastLogIndex)
+func (l *Log) append(e *LogEntry) error {
+	// Exit if log is not in a running state.
+	// Ignore replayed entries.
+	if l.state == Stopped {
+		return ErrClosed
+	} else if e.Index <= l.lastLogIndex {
```
Comment (otoolep) [inline raft/log.go]: Can this happen during normal processing?

File: raft/log.go [hunk 7]
```diff
@@ -1145,9 +1270,27 @@ func (l *Log) waitUncommitted(index uint64) error {
...
+		return nil
+	}
+
+	// If the entry is not the next then it may have changed leaders.
```
Comment (otoolep) [inline raft/log.go]: What may have changed leaders? Can you clarify the comment?

**Comment by otoolep** [comment]
I reviewed, most of my feedback is minor.

File: messaging/broker.go (hunk 1 above)
Comment (benbjohnson) [inline messaging/broker.go]: It's the number of bytes read but we're not supporting it. It's just an implementation of the [`io.ReaderFrom`](http://golang.org/pkg/io/#ReaderFrom) interface. We return an error for any partial reads.

File: raft/errors.go (hunk 2 above)
Comment (benbjohnson) [inline raft/errors.go]: No, the stream starts from the highest available index on the leader and the follower will trim its log or re-request a snapshot if needed.

File: raft/log.go (hunk 3 above)
Comment (benbjohnson) [inline raft/log.go]: Fixed: 947c776

File: raft/log.go (hunk 4 above)
Comment (benbjohnson) [inline raft/log.go]: Fixed: 947c776

File: raft/log.go (hunk 5 above)
Comment (benbjohnson) [inline raft/log.go]: I'd like to rework the `Wait()` so it doesn't poll so I'm not sure that it's worth making tickets.

File: raft/log.go (hunk 6 above)
Comment (benbjohnson) [inline raft/log.go]: Yes, since the applied index was moved to the FSM entirely, there's a small window where we can get duplicate indices immediately following a snapshot. We grab the applied index before making  the snapshot request from the FSM to ensure that we have all entries.

File: raft/log.go (hunk 7 above)
Comment (benbjohnson) [inline raft/log.go]: Fixed: 947c776
